include/cuco/detail/probing_scheme_impl.inl

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cuco {
+namespace experimental {
+namespace detail {
+/**
+ * @brief Probing iterator class.
+ *
+ * @tparam Extent Type of Extent
+ */
+template <typename Extent>
+class probing_iterator {
+ public:
+  using extent_type = Extent;                            ///< Extent type
+  using size_type   = typename extent_type::value_type;  ///< Size type
+
+  /**
+   *@brief Constructs an probing iterator
+   *
+   * @param start Iteration starting point
+   * @param step_size Double hashing step size
+   * @param upper_bound Upper bound of the iteration
+   */
+  __device__ constexpr probing_iterator(size_type start,
+                                        size_type step_size,
+                                        extent_type upper_bound) noexcept
+    : curr_index_{start}, step_size_{step_size}, upper_bound_{upper_bound}
+  {
+  }
+
+  /**
+   * @brief Dereference operator
+   *
+   * @return Current slot ndex
+   */
+  __device__ constexpr auto operator*() const noexcept { return curr_index_; }
+
+  /**
+   * @brief Prefix increment operator
+   *
+   * @return Current iterator
+   */
+  __device__ constexpr auto operator++() noexcept
+  {
+    // TODO: step_size_ can be a build time constant (e.g. linear probing)
+    //  Worth passing another extent type?
+    curr_index_ = (curr_index_ + step_size_) % upper_bound_;
+    return *this;
+  }
+
+  /**
+   * @brief Postfix increment operator
+   *
+   * @return Old iterator before increment
+   */
+  __device__ constexpr auto operator++(int32_t) noexcept
+  {
+    auto temp = *this;
+    ++(*this);
+    return temp;
```

Comment:
This might be a dumb question, but why do we need the `temp` here? Can't we just `return (*this)++`?

Reply:
> Can't we just return (*this)++?

Not sure I'm following. This will make postfix increment recursive.

Reply:
Right. I shouldn't do code reviews at 2am xD